lib/db/observed.go

```diff
@@ -0,0 +1,264 @@
+// Copyright (C) 2020 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package db
+
+import (
+	"time"
+
+	"github.com/syncthing/syncthing/lib/protocol"
+)
+
+func (db *Lowlevel) AddOrUpdatePendingDevice(device protocol.DeviceID, name, address string) error {
+	key := db.keyer.GeneratePendingDeviceKey(nil, device[:])
+	od := ObservedDevice{
+		Time:    time.Now().Round(time.Second),
+		Name:    name,
+		Address: address,
+	}
+	bs, err := od.Marshal()
+	if err == nil {
+		err = db.Put(key, bs)
+	}
+	if err != nil {
+		l.Warnf("Failed to store pending device entry: %v", err)
+	}
+	return err
+}
+
+// PendingDevices drops any invalid entries from the database after a
+// warning log message, as a side-effect.  That's the only possible
+// "repair" measure and appropriate for the importance of pending
+// entries.  They will come back soon if still relevant.
+func (db *Lowlevel) PendingDevices() (map[protocol.DeviceID]ObservedDevice, error) {
+	iter, err := db.NewPrefixIterator([]byte{KeyTypePendingDevice})
+	if err != nil {
+		return nil, err
+	}
+	defer iter.Release()
+	res := make(map[protocol.DeviceID]ObservedDevice)
+	for iter.Next() {
+		bs, err := db.Get(iter.Key())
+		if err != nil {
+			return nil, err
+		}
+		var od ObservedDevice
+		deviceID := db.keyer.DeviceFromPendingDeviceKey(iter.Key())
+		//FIXME: DeviceIDFromBytes() panics when given a wrong length input.
+		//       It should rather return an error which we'd check for here.
+		if len(deviceID) != protocol.DeviceIDLength {
+			goto deleteKey
+		}
+		if err := od.Unmarshal(bs); err != nil {
+			goto deleteKey
+		}
+		res[protocol.DeviceIDFromBytes(deviceID)] = od
+		continue
+	deleteKey:
+		l.Warnf("Invalid pending device entry, deleting from database: %x", iter.Key())
+		if err := db.Delete(iter.Key()); err != nil {
+			return nil, err
+		}
+	}
+	return res, nil
+}
+
+func (db *Lowlevel) AddOrUpdatePendingFolder(id, label string, device protocol.DeviceID) error {
+	key, err := db.keyer.GeneratePendingFolderKey(nil, device[:], []byte(id))
+	if err == nil {
+		of := ObservedFolder{
+			Time:  time.Now().Round(time.Second),
+			Label: label,
+		}
+		bs, err := of.Marshal()
+		if err == nil {
+			err = db.Put(key, bs)
+		}
+	}
+	if err != nil {
+		l.Warnf("Failed to store pending folder entry: %v", err)
+	}
+	return err
```

Comment:
Warn or return the error, not both.

Reply:
Sorry I missed this one (and the same in `AddOrUpdatePendingDevice()`).

In this case I think it makes sense to actually pass it up to the caller.  It will still only cause a warning message, as the API event should be generated anyway IMHO.

Reply:
Fixed (hopefully) in 441584c53e2be1b4aefa191a5b7cce20a8e10ee8.